### PR TITLE
Merging dev into master for release v0.7.0

### PR DESCRIPTION
**Pull request by gasse**

### CODE REVIEW

File: libecole/include/ecole/environment/environment.hpp [hunk 1]
```diff
@@ -15,6 +15,12 @@
 #include "ecole/scip/type.hpp"
 #include "ecole/traits.hpp"
 
+#include <iostream>
```
Comment (dchetelat) [inline libecole/include/ecole/environment/environment.hpp:18]: This should be removed (mistake)

File: python/src/ecole/environment.py [hunk 2]
```diff
@@ -109,16 +109,21 @@ def reset(self, instance, *dynamics_args, **dynamics_kwargs):
...
             done, action_set = self.dynamics.reset_dynamics(
                 self.model, *dynamics_args, **dynamics_kwargs
             )
 
-            observation = self.observation_function.extract(self.model, done)
+            if not done:
+                observation = self.observation_function.extract(self.model, done)
+            else:
+                observation = None
             reward_offset = self.reward_function.extract(self.model, done)
```
Comment (dchetelat) [inline python/src/ecole/environment.py:123]: Wasn't `reward_offset` supposed to be collected before the observation now? Maybe I messed up something when rebasing.

File: python/src/ecole/environment.py [hunk 3]
```diff
@@ -170,9 +175,15 @@ def step(self, action, *dynamics_args, **dynamics_kwargs):
             done, action_set = self.dynamics.step_dynamics(
                 self.model, action, *dynamics_args, **dynamics_kwargs
             )
-            observation = self.observation_function.extract(self.model, done)
+
+            if not done:
+                observation = self.observation_function.extract(self.model, done)
+            else:
+                observation = None
             reward = self.reward_function.extract(self.model, done)
```
Comment (dchetelat) [inline python/src/ecole/environment.py:183]: Same here.

File: libecole/include/ecole/environment/environment.hpp [hunk 4]
```diff
@@ -109,10 +116,14 @@ class Environment {
...
+
+			auto observation = OptionalObservation{};
+			if (!done) {
+				observation = observation_function().extract(model(), done);
+			}
 			return {
-				observation_function().extract(model(), done),
+				std::move(observation),
 				std::move(action_set),
 				reward_function().extract(model(), done),
```
Comment (dchetelat) [inline libecole/include/ecole/environment/environment.hpp:128]: I realize now also, should this be computed before the observation?

File: libecole/include/ecole/environment/environment.hpp [hunk 5]
```diff
@@ -153,16 +164,21 @@ class Environment {
...
 
+			auto observation = OptionalObservation{};
+			if (!done) {
+				observation = observation_function().extract(model(), done);
+			}
 			return {
-				observation_function().extract(model(), done),
+				std::move(observation),
 				std::move(action_set),
 				reward_function().extract(model(), done),
```
Comment (dchetelat) [inline libecole/include/ecole/environment/environment.hpp:183]: Same here.

**Comment by gasse** [comment]
@dchetelat yes that's normal, I forgot to merge #212 before releasing v0.7.0

The plan of action is the following:
1. we fast-forward `master` on `dev` (this PR), so that `master` is synced with the latest stable release, v0.7.0
2. we merge #212 into `dev`
3. we release a minor version v0.7.1 from `dev`
4. again, we fast-forward `master` on `dev`, so that `master` is synced with the latest stable release, v0.7.1

@AntoinePrv @dchetelat how does that sound ?

Ideally I would have done 1. right after I released v0.7.0, but I had to wait for one of you to approve the review (the PR was locked). This is a bit annoying, as while we are waiting the `stable` doc is not synced properly with the latest release @AntoinePrv do you think there is a way around it ?

Best,
Maxime

**Review by dchetelat — APPROVED** [review]

**Pull request merged**